add keylog support to OpenSSLContext object

schveiguy opened the pull request:
Ping @geod24

I'm not sure how to make sure this builds based on the openssl version. I think we need a release of deimos openssl.

schveiguy committed:
add keylog support to OpenSSLContext object

schveiguy commented:
See #8

Geod24 commented on tls/vibe/stream/openssl.d:
```diff
@@ -49,6 +49,10 @@ version(VibeForceALPN) enum alpn_forced = true;
 
+version(VibeForceKeylog) enum keylog_forced = true;
+else enum keylog_forced = false;
+enum haveKeylog = OPENSSL_VERSION_AT_LEAST(1, 1, 1) || keylog_forced;
```
IMO those should be `private`.

Geod24 commented on tls/vibe/stream/openssl.d:
```diff
@@ -849,6 +853,34 @@ final class OpenSSLContext : TLSContext {
+	@property void keylogCallback(SSL_CTX_keylog_cb_func cb)
+	{
+		logDebug("Keylog callback");
+		static if (haveKeylog) {
```
I would conditionally define the function, not have it do nothing. Also it would be good to have it have the same behavior as cURL, which is to pick up the environment variable by default.

Geod24 reviewed (commented):
Did the release, waiting for the registry to pick it up.

schveiguy committed:
Only allow keylog callback setting when keylog support is available.
Add mechanism to write each key to a keylog file based on the
SSLKEYLOGFILE environment variable similar to cURL.

schveiguy commented on tls/vibe/stream/openssl.d:
```diff
@@ -1225,6 +1254,35 @@ final class OpenSSLContext : TLSContext {
 
+static if (haveKeylog) {
+	/**
+	 * Utility callback to use the SSLKEYLOGFILE environment variable to log the key. Set on
```
Ugh, my connection crashed in the middle of editing a file, and apparently some of this didn't save.

schveiguy commented on tls/vibe/stream/openssl.d:
```diff
@@ -1225,6 +1254,35 @@ final class OpenSSLContext : TLSContext {
+			import core.file;
+			// if the file is not yet open, open it.
+			static FileStream keyfile;
+			if (!keyfile.isOpen) {
```
I guess we need a separate boolean gate.

schveiguy committed:
address review

schveiguy commented on tls/vibe/stream/openssl.d:
```diff
@@ -1225,6 +1254,35 @@ final class OpenSSLContext : TLSContext {
+		static extern(C) void callback(const SSL* ssl, const char* line) {
+			import core.file;
+			// if the file is not yet open, open it.
+			static FileStream keyfile;
```
Just realized, I need to close it too.

Geod24 replied:
Most likely the file will just be open for the duration of the program though.

schveiguy replied:
I have it figured out. I need to close it on thread exit.

schveiguy committed:
Clean up keylog file.

schveiguy committed:
Fix stupid bugs.

schveiguy committed:
Try a test where we log the key file.
Remove unnecessary pointer in keylogOnEnvVar
Make haveKeylog public so it can be used to determine whether to call
the functions.
Fix log file opening open mode.
Fake keylog file for platforms which don't support keylogging (to get
past test)

Geod24 commented on run-ci.sh:
```diff
@@ -29,8 +29,11 @@ else
 fi
 
 if [ ${RUN_TEST=1} -eq 1 ]; then
+    rm -f tests/loggedkeys.txt
```
I think rather than having it at this level, we should bring it lower, as we've seen it can have an impact on `dub`. Either nesting the invocation or running a preGenerate / postRun script ?

schveiguy replied:
Well, we can remove it directly in the code at that point.

schveiguy replied:
Done

Geod24 commented on tests/tls.d:
```diff
@@ -226,6 +227,24 @@ void testConn(TLSVersion cli_version, TLSVersion srv_version, bool expect_succes
 			}
+			// test logging the keyfile
+			import vibe.stream.openssl;
+			static if (haveKeylog) {
+				if (auto osslctx = cast(OpenSSLContext)cctx) {
+					osslctx.keylogOnEnvVar();
+				}
```
We shouldn't have to do this. It should be automatic if the variable is set.

schveiguy replied:
Given that we just realized the next version of openssl is going to have this behavior built in, do we really want to add this as default behavior now?

schveiguy replied:
I'll put the option behind a version.

schveiguy replied:
done.

Geod24 commented on tls/vibe/stream/openssl.d:
```diff
@@ -1225,6 +1250,51 @@ final class OpenSSLContext : TLSContext {
+	 * This must be called on every OpenSSLContext you wish to log the key. If the
+	 * environment variable is not set, this does not log the key.
+	 */
+	import vibe.core.file;
+	private FileStream keyfile;
+	void keylogOnEnvVar(OpenSSLContext context) {
+		static extern(C) void callback(const SSL* ssl, const char* line) {
+			if (line is null) return;
+
+			// use a separate boolean to avoid always checking for an env var when it doesn't exist.
+			static bool initialized;
+			if (!initialized) {
+				initialized = true;
+				// read the environment variable
+				import std.process;
+
+				import vibe.core.path;
+				auto path = NativePath(environment.get("SSLKEYLOGFILE", null));
+				if (path.empty) return;
+				// we use append like latest openssl does.
+				keyfile = openFile(path, FileMode.append);
+			}
+
+			if (!keyfile.isOpen) return;
+			keyfile.write(line[0 .. strlen(line)]);
+			keyfile.write("\n");
+		}
+
+		context.keylogCallback = &callback;
+	}
+
+	static ~this() {
+		keyfile.close();
+	}
```
Are we sure that this can never be called from a different thread ? The context would have to leak, obviously.

schveiguy replied:
keyfile is TLS. so a TLS static destructor seems like the right place to close it.

schveiguy replied:
It's now a `__gshared File` from phobos. Vibe was too annoying to do this well.

Geod24 commented on tls/vibe/stream/openssl.d:
```diff
@@ -1225,6 +1250,51 @@ final class OpenSSLContext : TLSContext {
+			keyfile.write("\n");
+		}
+
+		context.keylogCallback = &callback;
```
As mentioned this should be moved to the ctor

schveiguy replied:
Did not move it, but now calling it from the constructor (if the version is specified).

schveiguy committed:
Remove keylog file inside main instead of in CI script.

schveiguy committed:
Switch keylogfile to Phobos. Set up a version specifier that can enable
the keylogfile behavior by default.

Geod24 commented on run-ci.sh:
```diff
@@ -33,4 +33,6 @@ if [ ${RUN_TEST=1} -eq 1 ]; then
         (cd tests && dub --temp-build --single $ex $DUB_FLAGS)
     done
+    # check the keylog file was written
+    test -f "tests/loggedkeys.txt"
```
I would also move this to the test ?

schveiguy replied:
Don't we care that the file is written after the program exits?

schveiguy replied:
Moved to the main function in the test.

Geod24 commented on tests/tls.d:
```diff
@@ -226,6 +228,20 @@ void testConn(TLSVersion cli_version, TLSVersion srv_version, bool expect_succes
 				stunnel.finalize();
 				return;
 			}
+			// test logging the keyfile
+			import vibe.stream.openssl;
+			static if (!haveKeylog) {
+				// still write the keylog file, so the test can pass
+				import std.process;
+				import vibe.core.path;
+				auto path = NativePath(environment.get("SSLKEYLOGFILE", null));
+				if (!path.empty) {
+					import vibe.core.file;
+					auto keyfile = openFile(path, FileMode.append);
+					// no need to write anything, the file just needs to exist.
+					keyfile.close();
+				}
+			}
```
This shouldn't be needed anymore ?

Geod24 commented on tls/vibe/stream/openssl.d:
```diff
@@ -49,6 +49,10 @@ version(VibeForceALPN) enum alpn_forced = true;
 else enum alpn_forced = false;
 enum haveALPN = OPENSSL_VERSION_NUMBER >= 0x10200000 || alpn_forced;
 
+version(VibeForceKeylog) enum keylog_forced = true;
```
I would remove this `VibeForceKeylog` and instead twek it so that `VibeKeylogFromEnvironment` is taken into account. This way you don't have to have two checks on L802 (in the ctor).

Geod24 replied:
Or just remove this part and have the `version` in ctor

schveiguy replied:
ok

schveiguy replied:
We need the version in the ctor otherwise, this code is on by default.

schveiguy replied:
I removed the keylog_forced thing, but still have the version for the env var reading.

Geod24 commented on tls/vibe/stream/openssl.d:
```diff
@@ -849,6 +858,27 @@ final class OpenSSLContext : TLSContext {
 	/// Get the current ALPN callback function
 	@property TLSALPNCallback alpnCallback() const { return m_alpnCallback; }
 
+	static if (haveKeylog) {
```
```suggestion
	static if (OPENSSL_VERSION_AT_LEAST(1, 1, 1)) {
```

I think that would be enough.

schveiguy replied:
this removes the point of the "force" version.

I'm OK removing that though. Maybe I shouldn't have copied that in the first place.

Geod24 commented on tls/vibe/stream/openssl.d:
```diff
@@ -795,6 +799,11 @@ final class OpenSSLContext : TLSContext {
 		if (kind == TLSContextKind.client) peerValidationMode = TLSPeerValidationMode.trustedCert;
 		else peerValidationMode = TLSPeerValidationMode.none;
 
+		version(VibeKeylogFromEnvironment) static if(haveKeylog) {
+			// automatically set up the SSLKEYLOGFILE environment handling.
+			() @trusted {this.keylogOnEnvVar();} ();
+		}
```
```suggestion
		version (VibeKeylogFromEnvironment) {
			// automatically set up the SSLKEYLOGFILE environment handling.
			if (auto path = environment.get("SSLKEYLOGFILE", null))
				() @trusted {this.keylogOnEnvVar();} ();
		}
```

This way there is less overhead if the env variable isn't set.

schveiguy replied:
I don't like having 2 places that use the same check. I can move this to the function though. I have a thought on how to make this better.

schveiguy committed:
Tweaks for review.

Geod24 commented on tls/vibe/stream/openssl.d:
```diff
@@ -1225,6 +1253,59 @@ final class OpenSSLContext : TLSContext {
+	 * This must be called on every OpenSSLContext you wish to log the key. If the
+	 * environment variable is not set, this does not log the key.
+	 */
+	void keylogOnEnvVar(OpenSSLContext context) {
```
```suggestion
	void setSSLKEYLOGFILECallback(OpenSSLContext context) {	
```
```suggestion
	void setSSLKeyLogFileCallback(OpenSSLContext context) {	
```

Or something similar ? The environment variable name is well-known at this point (and was pioneered by cURL AFAIK, OpenSSL just copied it), so I think the first version is the least ambiguous, but not very in line with the common style.
Another option is to use `install` instead of `set`.

Geod24 commented on tls/vibe/stream/openssl.d:
```diff
@@ -1225,6 +1253,59 @@ final class OpenSSLContext : TLSContext {
+	import std.stdio;
+	private __gshared File keyfile;
+	/**
+	 * Use the SSLKEYLOGFILE environment variable to log the key.
```
```suggestion
	 * Set up OpenSSL to output master-secret keys to SSLKEYLOGFILE
```

Geod24 commented on tls/vibe/stream/openssl.d:
```diff
@@ -1225,6 +1253,59 @@ final class OpenSSLContext : TLSContext {
+	 * determines where the keys will be logged. Each key will be logged on a
+	 * separate line.
+	 *
+	 * This must be called on every OpenSSLContext you wish to log the key. If the
+	 * environment variable is not set, this does not log the key.
```
```suggestion
	 * This must be called on every OpenSSLContext you wish to log the key. If the
	 * environment variable is not set, this does nothing.
	 *
	 * See_Also: https://everything.curl.dev/usingcurl/tls/sslkeylogfile.html
```

Geod24 commented on tls/vibe/stream/openssl.d:
```diff
@@ -1225,6 +1253,59 @@ final class OpenSSLContext : TLSContext {
+			if (!initialized) {
+				initialized = true;
+				// read the environment variable
+				import std.process;
```
Stray import

Geod24 commented on tls/vibe/stream/openssl.d:
```diff
@@ -49,6 +49,8 @@ version(VibeForceALPN) enum alpn_forced = true;
 else enum alpn_forced = false;
 enum haveALPN = OPENSSL_VERSION_NUMBER >= 0x10200000 || alpn_forced;
 
+enum haveKeylog = OPENSSL_VERSION_AT_LEAST(1, 1, 1);
```
`private` ?

schveiguy replied:
Since we optionally define the functions based on this, leaving the enum public helps determine whether we can register the callback (in the event where we don't want to use the env var, but directly assign the callback).

Otherwise, one has to use `__traits(compiles, ...)` or `is(typeof(...))`

Geod24 replied:
Yes but there's no currently-supported version of OpenSSL that would not have this. OpenSSL 1.1.1 was released in 2018 and LTS are 5 years for OpenSSL. It's highly unlikely someone would need a `__traits`.

schveiguy replied:
CI says otherwise, as I had to add static ifs for this in the test (or it wouldn't compile).

Geod24 commented on tls/vibe/stream/openssl.d:
```diff
@@ -795,6 +797,11 @@ final class OpenSSLContext : TLSContext {
 		if (kind == TLSContextKind.client) peerValidationMode = TLSPeerValidationMode.trustedCert;
 		else peerValidationMode = TLSPeerValidationMode.none;
 
+		version(VibeKeylogFromEnvironment) static if(haveKeylog) {
```
```suggestion
		version (VibeKeylogFromEnvironment) static if (haveKeylog) {
```

schveiguy replied:
version spacing in the file does not have a separating space. Static if does. So I fixed that.

Geod24 commented on tests/tls.d:
```diff
@@ -282,6 +284,18 @@ void testVersion()
 {
+	// check that the loggedkeys.txt file exists
+	import vibe.stream.openssl;
+	static if(haveKeylog) {
```
```suggestion
	static if (haveKeylog) {
```

Geod24 reviewed (approved):
LGTM. Made a few comment, most important one about naming the public function.

Geod24 commented:
@s-ludwig : Would you like to comment on the feature ? I'm certain this will be quite useful for developers, having used it extensively myself, but LMK if you disagree. Otherwise once @schveiguy is done with the last nits I will pull this.

schveiguy committed:
Redo file management for keylog file, make it actually thread safe. Fix
review spacing problems.

schveiguy committed:
add curl link to docs.

Co-authored-by: Mathias LANG <geod24@gmail.com>

schveiguy committed:
More review fixes. Fix synchronization for keyfile opening.

schveiguy commented:
Take note @Geod24 I realized the synchronization was very wrong. So I encapsulated the file opening into a struct with a mutex (note, openssl has a mutex for this too).

Geod24 commented on tls/vibe/stream/openssl.d:
```diff
@@ -1254,9 +1254,60 @@ final class OpenSSLContext : TLSContext {
+		private File keyfile;
+		private TaskMutex mutex;
+
+		File initialize() shared {
+			mutex.lock();
+			scope(exit) mutex.unlock();
+			return (cast(KeylogFile*)&this).initializeImpl();
+		}
```
https://dlang.org/library/std/concurrency/init_once.html ?

schveiguy replied:
Unaware of this! I was looking in typecons.

Commit:
Flush stream after every line output to work better with wireshark.

s-ludwig commented:
> @s-ludwig : Would you like to comment on the feature ? I'm certain this will be quite useful for developers, having used it extensively myself, but LMK if you disagree. Otherwise once @schveiguy is done with the last nits I will pull this.

I haven't been able to follow this, but looks good to me, considering its behind `version (VibeKeylogFromEnvironment)`. It would just be nice to have a short explanation in the `version` table of the docs: https://github.com/vibe-d/vibed.org/blob/6ce90b1471808ae7283a3181f1f34584eb71dc22/views/docs.dt#L978
